src/accessToken.ts

```diff
@@ -0,0 +1,222 @@
+import { getClaims, type JWTPayload } from "@workos-inc/authkit-js";
+import { useCallback, useEffect, useMemo, useReducer, useRef } from "react";
+import { useAuth } from "./hook";
+
+interface TokenState {
+  token: string | undefined;
+  loading: boolean;
+  error: Error | null;
+}
+
+type TokenAction =
+  | { type: "FETCH_START" }
+  | { type: "FETCH_SUCCESS"; token: string | undefined }
+  | { type: "FETCH_ERROR"; error: Error }
+  | { type: "RESET" };
+
+function tokenReducer(state: TokenState, action: TokenAction): TokenState {
+  switch (action.type) {
+    case "FETCH_START":
+      return { ...state, loading: true, error: null };
+    case "FETCH_SUCCESS":
+      return { ...state, loading: false, token: action.token };
+    case "FETCH_ERROR":
+      return { ...state, loading: false, error: action.error };
+    case "RESET":
+      return { ...state, token: undefined, loading: false, error: null };
+    // istanbul ignore next
+    default:
+      return state;
+  }
+}
+
+const TOKEN_EXPIRY_BUFFER_SECONDS = 60;
+const MIN_REFRESH_DELAY_SECONDS = 15; // minimum delay before refreshing token
+const RETRY_DELAY_SECONDS = 300; // 5 minutes
+
+interface TokenData {
+  exp: number;
+  timeUntilExpiry: number;
+  isExpiring: boolean;
+}
+
+function parseToken(token: string): TokenData | null {
+  try {
+    const claims = getClaims(token);
+    const now = Date.now() / 1000;
+    const exp = claims.exp ?? 0;
+    const timeUntilExpiry = exp - now;
+    const isExpiring = timeUntilExpiry <= TOKEN_EXPIRY_BUFFER_SECONDS;
+
+    return { exp, timeUntilExpiry, isExpiring };
+  } catch {
+    return null;
+  }
+}
+
+function getRefreshDelay(timeUntilExpiry: number): number {
+  const refreshTime = Math.max(
+    timeUntilExpiry - TOKEN_EXPIRY_BUFFER_SECONDS,
+    MIN_REFRESH_DELAY_SECONDS,
+  );
+  return refreshTime * 1000; // convert to milliseconds
+}
+
+/**
+ * A hook that manages access tokens with automatic refresh.
+ *
+ * @example
+ * ```ts
+ * const { accessToken, loading, error, refresh } = useAccessToken();
+ * ```
+ *
+ * @returns An object containing the access token, loading state, error state, and a refresh function.
+ */
+export function useAccessToken() {
```

Comment:
the traditional way to get claims in authkit-react is via the `useAuth` hook. i think we should just extend that to support custom claims instead of bolting on a new hook.